File: app/src/main/java/com/scop/org/minesweeper/generators/RandomCheckedTestGenerator.java
package com.scop.org.minesweeper.generators;

import com.scop.org.minesweeper.control.MainLogic;
import com.scop.org.minesweeper.control.Settings;
import com.scop.org.minesweeper.elements.Grid;
import com.scop.org.minesweeper.elements.Tile;
import com.scop.org.minesweeper.generators.solver.BasicSolver;
import com.scop.org.minesweeper.generators.solver.DeepSolver;
import com.scop.org.minesweeper.generators.solver.PossibleSolver;
import com.scop.org.minesweeper.generators.solver.Sketch;
import com.scop.org.minesweeper.generators.solver.Solver;
import com.scop.org.minesweeper.utils.GridUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Set;

public class RandomCheckedTestGenerator extends RandomCheckedGenerator {

	@Override
	public void generateNewGrid(final Grid grid, int bombs, FinishCallback cb) {
		new Thread(){
			@Override
			public void run(){
				boolean test = false;


				if (test) {
					test1(grid, bombs);
					selectedSafeTile = 24;

					solve(grid);


				} else {
					generateNewRandomGrid(grid, bombs);
					selectedSafeTile = GridUtils.findSafeOpenTileIdx(grid);


					int numBombsLeft = 0;

					while (numBombsLeft == 0) {
						Set<Integer> bombsLeft = solve(grid);
						numBombsLeft = bombsLeft == null ? 0 : bombsLeft.size();

						if (numBombsLeft == 0) {
							generateNewRandomGrid(grid, bombs);
							selectedSafeTile = GridUtils.findSafeOpenTileIdx(grid);
						}
					}

				}


				cb.finished();
			}
		}.start();
	}

	protected void test1(Grid grid, int bombs) {
		int w = grid.getW();
		int h = grid.getH();
		List<Tile> tiles = grid.getGrid();
		tiles.clear();


		for (int j=0; j<h; j++) {
			for (int i=0; i<w; i++) {
				tiles.add( new Tile(i,j,Tile.Status.COVERED) );
			}
		}


		for (int idx : new int[]{8,10,11,12,17,22}) {
			tiles.get(idx).plantBomb();

			GridUtils.getNeighbors(grid, tiles.get(idx)).forEach(Tile::hasBombNear);
		}
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/RandomCheckedTestGenerator.kt
package org.oar.minesweeper.generators

import org.oar.minesweeper.elements.Grid
import org.oar.minesweeper.elements.GridStartOptions
import org.oar.minesweeper.elements.Tile
import org.oar.minesweeper.elements.TileStatus
import org.oar.minesweeper.utils.GridUtils.findSafeOpenTileIdx
import org.oar.minesweeper.utils.GridUtils.getNeighbors

class RandomCheckedTestGenerator : RandomCheckedGenerator() {
    override fun generateNewGrid(grid: Grid, onFinish: (GridStartOptions) -> Unit) {
        Thread {
            val test = false
            if (test) {
                test1(grid)
                selectedSafeTile = 24
                solve(grid)

            } else {
                generateNewRandomGrid(grid)
                selectedSafeTile = grid.findSafeOpenTileIdx()

                var numBombsLeft = 0

                while (numBombsLeft == 0) {
                    val bombsLeft = solve(grid)
                    numBombsLeft = bombsLeft?.size ?: 0

                    if (numBombsLeft == 0) {
                        generateNewRandomGrid(grid)
                        selectedSafeTile = grid.findSafeOpenTileIdx()
                    }
                }
            }
            onFinish(
                GridStartOptions(selectedSafeTile)
            )
        }.start()
    }

    private fun test1(grid: Grid) {
        val w = grid.width
        val h = grid.height

        val tiles = grid.tiles
        tiles.clear()
        for (j in 0 until h) {
            for (i in 0 until w) {
                tiles.add(Tile(i, j, TileStatus.COVERED))
            }
        }
        for (idx in intArrayOf(8, 10, 11, 12, 17, 22)) {
            tiles[idx].hasBomb = true
            grid.getNeighbors(tiles[idx]).forEach { it.hasBombNear() }
        }
    }
}